torch/csrc/jit/mobile/versioned_operators.cpp

```diff
@@ -0,0 +1,185 @@
+// bytecode_version_in_model), check the compatibility for that operator. If
+// it's not compatible, throw an error with reason. If it's compatible, do
+// compatibility treatments if necessary and return the corresponding operator
+// functor that is compatible to this runtime. The logic is implemented in
+// function operator_resolver().
+//
+// Under the hood, a static table,
+// op_version_table, is saved for current runtime. The keys are the operator
+// names and values are a container with version numbers and their metadata. The
+// incoming operator with version is searched in the table to find the match.
+// Note that the size of op_version_table is small because it only contains
+// operators that have more than one versions. For most of the operators, they
+// fall through to the existing implementation with default version 0.
+
+namespace torch {
+namespace jit {
+namespace mobile {
+namespace {
+
+// The version info is limited to combinations of two situations:
+//   * Route to another op name
+//   * Push additional default values
+// It’s fine to update the content of version info, to handle more situations in
+// fugure. Since it’s implementation details, it should not affect
+// compatibility.
+struct VersionInfo {
+  VersionInfo(
+      int64_t version,
+      const c10::OperatorName& name,
+      c10::optional<IValue> default_val)
+      : version(version), name(name), default_val(default_val) {}
+  int64_t version;
+  c10::OperatorName name;
+  c10::optional<IValue> default_val;
+};
+
+struct VersionInfoContainer {
+  VersionInfoContainer(const std::vector<VersionInfo>& version_vec)
+      : version_vec(version_vec) {}
+  int64_t min_version() const {
+    return version_vec.begin()->version;
+  }
+
+  int64_t max_version() const {
+    return version_vec.rbegin()->version + 1;
+  }
+
+  std::vector<VersionInfo> version_vec;
+};
+
+VersionInfo getVersioninfo(
+    const VersionInfoContainer& container,
+    const c10::OperatorName& name,
+    int64_t op_version) {
+  // Backward compatibility breakage
+  TORCH_CHECK(
+      op_version >= container.min_version(),
+      "The version number of ",
+      c10::toString(name),
+      ", ",
+      op_version,
+      ", is smaller than the minimum version number, ",
+      container.min_version(),
+      ", supported in this runtime. The model file is too old. ",
+      "Update the model with most recent code.");
+
+  // Forward compatibility breakage
+  TORCH_CHECK(
+      op_version <= container.max_version(),
+      "The version number of ",
+      c10::toString(name),
+      ", ",
+      op_version,
+      ", is larger than the maximum version number, ",
+      container.max_version(),
+      ", supported in this runtime. Update the runtime to be compatible",
+      " to this model.");
+
+  for (const auto& info : container.version_vec) {
+    if (info.version == op_version) {
+      return info;
+    }
+  }
+
+  if (op_version == container.max_version() + 1) {
```

Comment:
Hmm why current version is max_version() + 1?

Reply:
Good catch! There should be no "+1" here.

Reply:
This is interesting. So if an operation, like convolution, changes, then it will require "dummy" version infos forever so that something in the vector (or at the end of the vector) matches these checks? It also seems like operators that have had multiple changes won't be supported (although this framework could be extended later). 

For example, if there's a function foo(), and over time two arguments are added to foo(), how would that work? So let's say that version 1 of the runtime adds an argument 'a' to foo(), and version 3 of the runtime adds an argument 'b' to foo(). With this approach I don't think foo() from version 0 can be run on version 3 or later, right? And foo() would require an entry for version 1 that adds 'b', an entry for version 2 that adds 'b' and no entry for version 3 as long as the runtime version is also 3. But when the runtime version becomes 4 then foo() will also need an entry for version 3, but that entry won't do anything?

Is that correct?

Reply:
Very good point! The table can only support one version that add one arg. If there is another version adding another arg, it won't work. One option is to have a vector of default args (IValues) to the operator version table. In your example, when runtime is on v1, the table would be like {"foo": {0: ["a"]}}. When runtime is on v2, the table is {"foo": {0: ["a"]}, {1: []}} (neglecting changes other than adding args). When runtime is on v3, the table is {"foo": {0: ["a", "b"]}, {1: ["b"]}, {2: ["b"]}}. Would this work?

Reply:
It might be interesting to consider an approach like the operator versioning implemented for torchscript. You could, for example, have each op entry have a start version and end version. That way when loading from version X, you would scan through the former versions and find which op entry X is in, or not discover it and know that no additional processing was needed. 

Sticking with only being able to add a single arg seems OK for now. If you need more advanced conversions in the future I think you can write them later.

```
struct VersionInfo {
  VersionInfo(
    const int64_t _start_version,
    const int64_t _end_version,
    const c10::OperatorName& _name,
    const c10::optional<IValue>& _default_val) 
      : start_version_{_start_version}, end_version_{_end_version}, name_{_name}, default_val_{_default_val} { }

  ...
};
```

So when getting an operator foo with version X, the code can enumerate the VersionInfos like today but check their start and end versions:

```
for (const auto& vi : version_infos) {
  if (version >= vi.start_version_ && version <= vi.end_version_) {
    // match!
    ...
  }
}

// no match found! that means no compatibility is needed and the op can be run normally
```

What do you think of an approach like this? I think it might simplify things.

Reply:
I get the point of keeping _start_version and _end_version in each entry. It's especially useful for torchscript versioning, where the op version is bond to the global model version. In such case, even there's no change in op_A, but since op_B is changed, we need to bump _end_version of op_A as well. 

In this PR we decoupled the global model version, and make op version unique for each op. In such case, each entry must be different for different op version. i.e., if op_A is compatibility break, we must add another entry with new version to op_A, otherwise it's not "break". Keeping _start_version and _end_version in each entry may not be practical. 

And out range (no match in your example) does not necessarily mean the op can be run normally. If the op version X is too old that we decide from some time we will not support versions older than X for this op, we should throw and error instead of run it normally.

Reply:
I'm not sure I follow. Maybe a VC would be more helpful than Github to discuss this issue.

What I don't understand, in particular, is:

- I'm not sure why start and end can't refer to op versions
- if an Op from version X shouldn't be run then it seems like that could be modeled as a 0 start X end entry that says to throw a runtime error

Reply:
Yeah a VC would make it more efficient 😄 
I'm not saying start and end can't refer to op versions. Just saying they may not be necessary because each time we bump a version is for a breaking reason. i.e. start == end